sw/device/tests/dif_gpio/mock_mmio.h

```diff
@@ -0,0 +1,59 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+extern "C" {
+#include "sw/device/lib/base/mmio.h"
+}
+
+// Class for mocking |reg32_read()| and |reg32_write()|.
+// This class should only be used by MmioMockTest and
+// the private constructor expresses this expectation.
+class MmioMock {
+ public:
+  MOCK_METHOD(uint32_t, reg32_read, (reg32_t, ptrdiff_t));
+  MOCK_METHOD(void, reg32_write, (reg32_t, ptrdiff_t, uint32_t));
+
+ private:
+  MmioMock() = default;
+
+  friend class MmioMockTest;
+};
+
+// Fixture for constructing and destroying a static MmioMock instance
+// that can be used from |reg32_read()| and |reg32_write()| mocks.
+class MmioMockTest : public testing::Test {
+ protected:
+  void SetUp() override;
+  void TearDown() override;
+
+  // Helper methods to simplify tests. They return a reference to the
+  // expectation so that they can be modified. Defined here to be able to use
+  // auto in the return type.
+  auto &EXPECT_REG32_WRITE(const reg32_t addr, const uint32_t offset,
+                           const uint32_t val) {
+    return EXPECT_CALL(*mock_, reg32_write(addr, offset, val))
+        .InSequence(expectationSequence_);
+  }
+
+  auto &EXPECT_REG32_READ(const reg32_t addr, const uint32_t offset) {
+    return EXPECT_CALL(*mock_, reg32_read(addr, offset))
+        .InSequence(expectationSequence_);
+  }
+
+ private:
+  testing::Sequence expectationSequence_;
+
+  // Static mock instance
+  static std::unique_ptr<MmioMock> mock_;
+
+  // Friends to be able to redispatch calls to MmioMock
+  friend uint32_t reg32_read(reg32_t, ptrdiff_t);
+  friend void reg32_write(reg32_t, ptrdiff_t, uint32_t);
```

Comment:
I'd avoid friends if possible. I don't think that there's anything unreasonable about making the mock functions public.